diff -Naur kmail-22.12.3/src/configuredialog/configureappearancepage.cpp kmail-22.12.3-patched/src/configuredialog/configureappearancepage.cpp
--- kmail-22.12.3/src/configuredialog/configureappearancepage.cpp	2023-02-27 05:02:12.000000000 +0100
+++ kmail-22.12.3-patched/src/configuredialog/configureappearancepage.cpp	2023-08-03 18:16:33.561676026 +0200
@@ -839,6 +839,12 @@
     systrayBoxlayout->addWidget(mShowNumberInTaskBar);
     connect(mShowNumberInTaskBar, &QCheckBox::stateChanged, this, &ConfigModuleTab::slotEmitChanged);
 
+    mSystemTrayShowUnreadMail = new QCheckBox(i18n("Show unread mail in tray icon"), this);
+    systrayBoxlayout->addWidget(mSystemTrayShowUnreadMail);
+    mSystemTrayShowUnreadMail->setEnabled(false);
+    connect(mSystemTrayShowUnreadMail, &QCheckBox::stateChanged, this, &ConfigModuleTab::slotEmitChanged);
+    connect(mSystemTrayCheck, &QAbstractButton::toggled, mSystemTrayShowUnreadMail, &QWidget::setEnabled);
+
     topLayout->addStretch(100); // spacer
 }
 
@@ -850,6 +856,7 @@
 {
     loadWidget(mSystemTrayCheck, KMailSettings::self()->systemTrayEnabledItem());
     loadWidget(mStartInTrayCheck, KMailSettings::self()->startInTrayItem());
+    loadWidget(mSystemTrayShowUnreadMail, KMailSettings::self()->systemTrayShowUnreadItem());
     loadWidget(mShowNumberInTaskBar, KMailSettings::self()->showUnreadInTaskbarItem());
     loadWidget(mCloseAfterReplyOrForwardCheck, MessageViewer::MessageViewerSettings::self()->closeAfterReplyOrForwardItem());
     mViewerSettings->readConfig();
@@ -858,6 +865,7 @@
 void AppearancePageGeneralTab::save()
 {
     saveCheckBox(mSystemTrayCheck, KMailSettings::self()->systemTrayEnabledItem());
+    saveCheckBox(mSystemTrayShowUnreadMail, KMailSettings::self()->systemTrayShowUnreadItem());
     saveCheckBox(mStartInTrayCheck, KMailSettings::self()->startInTrayItem());
     saveCheckBox(mShowNumberInTaskBar, KMailSettings::self()->showUnreadInTaskbarItem());
     KMailSettings::self()->save();
diff -Naur kmail-22.12.3/src/configuredialog/configureappearancepage.h kmail-22.12.3-patched/src/configuredialog/configureappearancepage.h
--- kmail-22.12.3/src/configuredialog/configureappearancepage.h	2023-02-27 05:02:12.000000000 +0100
+++ kmail-22.12.3-patched/src/configuredialog/configureappearancepage.h	2023-08-03 18:16:35.973706742 +0200
@@ -162,6 +162,7 @@
     QCheckBox *mSystemTrayCheck = nullptr;
     QCheckBox *mStartInTrayCheck = nullptr;
     QCheckBox *mShowNumberInTaskBar = nullptr;
+    QCheckBox *mSystemTrayShowUnreadMail = nullptr;
 };
 
 class TagListWidgetItem : public QListWidgetItem
diff -Naur kmail-22.12.3/src/kmsystemtray.cpp kmail-22.12.3-patched/src/kmsystemtray.cpp
--- kmail-22.12.3/src/kmsystemtray.cpp	2023-02-27 05:02:12.000000000 +0100
+++ kmail-22.12.3-patched/src/kmsystemtray.cpp	2023-08-03 18:16:38.457738376 +0200
@@ -20,6 +20,8 @@
 #include <QMenu>
 
 #include "widgets/kactionmenutransport.h"
+#include <QPainter>
+#include <KIconLoader>
 
 using namespace MailCommon;
 
@@ -42,6 +44,7 @@
     setToolTipTitle(i18n("KMail"));
     setToolTipIconByName(QStringLiteral("kmail"));
     setIconByName(QStringLiteral("kmail"));
+    mIcon=QIcon::fromTheme(QStringLiteral("mail-mark-unread-new"));
 
     KMMainWidget *mainWidget = kmkernel->getKMMainWidget();
     if (mainWidget) {
@@ -120,17 +123,76 @@
     updateToolTip(count);
 }
 
+void KMSystemTray::setShowUnreadCount(bool showUnreadCount)
+{
+    if (mShowUnreadMailCount == showUnreadCount) {
+        return;
+    }
+    mShowUnreadMailCount = showUnreadCount;
+    mUnityServiceManager->updateSystemTray();
+}
+
+void KMSystemTray::slotGeneralPaletteChanged()
+{
+    const KColorScheme scheme(QPalette::Active, KColorScheme::View);
+    mTextColor = scheme.foreground(KColorScheme::LinkText).color();
+    mUnityServiceManager->updateSystemTray();
+}
+
 /**
  * Update the count of unread messages.  If there are unread messages,
- * show the "unread new mail" KMail icon.
+ * overlay the count on top of a transparent version of the KMail icon.
  * If there is no unread mail, restore the normal KMail icon.
  */
 void KMSystemTray::updateCount(int count)
 {
     if (count == 0) {
         setIconByName(QStringLiteral("kmail"));
+        return;
+    }
+    if (mShowUnreadMailCount) {
+        const int overlaySize = KIconLoader::global()->currentSize(KIconLoader::Panel);
+
+        const QString countString = QString::number(count);
+        QFont countFont = QFontDatabase::systemFont(QFontDatabase::GeneralFont);
+        countFont.setBold(true);
+
+        // We now adjust the font size to fit the number inside the
+        // available space. We pick always the same size for numbers up to
+        // 99, and then decrease as needed.
+        // Remark:overlaySize is 32 pixels and the final icon is scaled
+        // down to 22 pixels, so the "natural" font size is meaningless anyway.
+        float countFontSize = countFont.pointSizeF();
+        QFontMetrics qfm(countFont);
+        const int width = qfm.horizontalAdvance(count < 100 ? QStringLiteral("99") : countString);
+        countFontSize *= float(overlaySize - 2) / float(width);
+        countFont.setPointSizeF(countFontSize);
+
+        // Paint the number in a pixmap
+        QPixmap overlayPixmap(overlaySize, overlaySize);
+        overlayPixmap.fill(Qt::transparent);
+
+        QPainter p(&overlayPixmap);
+        p.setFont(countFont);
+        if (!mTextColor.isValid()) {
+            slotGeneralPaletteChanged();
+        }
+
+        p.setBrush(Qt::NoBrush);
+        p.setPen(mTextColor);
+        p.setOpacity(1.0);
+        p.drawText(overlayPixmap.rect(), Qt::AlignCenter, countString);
+        p.end();
+
+        QPixmap iconPixmap = mIcon.pixmap(overlaySize, overlaySize);
+
+        QPainter pp(&iconPixmap);
+        pp.drawPixmap(0, 0, overlayPixmap);
+        pp.end();
+
+        setIconByPixmap(iconPixmap);
     } else {
-        setIconByName(QStringLiteral("mail-mark-unread-new"));
+        setIconByPixmap(mIcon);
     }
 }
 
diff -Naur kmail-22.12.3/src/kmsystemtray.h kmail-22.12.3-patched/src/kmsystemtray.h
--- kmail-22.12.3/src/kmsystemtray.h	2023-02-27 05:02:12.000000000 +0100
+++ kmail-22.12.3-patched/src/kmsystemtray.h	2023-08-03 18:16:40.677766648 +0200
@@ -30,6 +30,7 @@
     /** destructor */
     ~KMSystemTray() override;
 
+    void setShowUnreadCount(bool showUnreadCount);
     void hideKMail();
 
     void updateStatus(int count);
@@ -43,6 +44,9 @@
     void slotContextMenuAboutToShow();
     void slotSelectCollection(QAction *act);
 
+    void slotGeneralPaletteChanged();
+    void slotGeneralFontChanged();
+
     Q_REQUIRED_RESULT bool buildPopupMenu();
     void fillFoldersMenu(QMenu *menu, const QAbstractItemModel *model, const QString &parentName = QString(), const QModelIndex &parentIndex = QModelIndex());
     int mDesktopOfMainWin = 0;
@@ -51,6 +55,10 @@
     bool mHasUnreadMessage = false;
     bool mIconNotificationsEnabled = true;
 
+    QColor mTextColor;
+    QIcon mIcon;
+    bool mShowUnreadMailCount = true;
+
     QMenu *mNewMessagesPopup = nullptr;
     QAction *mSendQueued = nullptr;
     KMail::UnityServiceManager *mUnityServiceManager = nullptr;
diff -Naur kmail-22.12.3/src/settings/kmail.kcfg.cmake kmail-22.12.3-patched/src/settings/kmail.kcfg.cmake
--- kmail-22.12.3/src/settings/kmail.kcfg.cmake	2023-02-27 05:02:12.000000000 +0100
+++ kmail-22.12.3-patched/src/settings/kmail.kcfg.cmake	2023-08-03 18:16:43.169798382 +0200
@@ -89,6 +89,10 @@
         <label>Enable system tray icon</label>
         <default>false</default>
       </entry>
+      <entry name="SystemTrayShowUnread" type="Bool">
+        <label>Show unread mail in system tray</label>
+        <default>true</default>
+      </entry>
       <entry name="StartInTray" type="Bool">
         <label>Start minimized to tray</label>
         <default>false</default>
diff -Naur kmail-22.12.3/src/unityservicemanager.cpp kmail-22.12.3-patched/src/unityservicemanager.cpp
--- kmail-22.12.3/src/unityservicemanager.cpp	2023-02-27 05:02:12.000000000 +0100
+++ kmail-22.12.3-patched/src/unityservicemanager.cpp	2023-08-03 18:16:46.909846008 +0200
@@ -230,5 +230,8 @@
             delete mSystemTray;
             mSystemTray = nullptr;
         }
+        if (mSystemTray) {
+            mSystemTray->setShowUnreadCount(KMailSettings::self()->systemTrayShowUnread());
+        }
     }
 }
